persia/ctx.py

```diff
@@ -208,102 +327,8 @@ def prepare_features(self, batch, is_training=True):
-                # no duplicate id in raw_id_tensor
-                (
-                    raw_embedding,
-                    index,
-                    non_empty_index,
-                    sample_id_num,
-                ) = emb.get_raw_embedding()
-
-                batch.emb_slot.append([raw_embedding, index, non_empty_index])
-
-                distinct_id_tensor = pte.ptr_to_tensor_f16(
-                    raw_embedding.data_ptr(), raw_embedding.shape(), False
-                )
-                index_tensor = pte.ptr_to_tensor_long(
-                    index.data_ptr(),
-                    index.shape(),
-                )  # tensor shape (1, batch_size * sample_fixed_size)
-                max_index = index_tensor.max()
-                size_of_distinct_id_tensor = distinct_id_tensor.shape[0]
-                torch.cuda.synchronize()
-
-                assert (
-                    max_index < size_of_distinct_id_tensor
-                ), "raw embedding select index larger than tensor"
-                non_empty_index_tensor = pte.ptr_to_tensor_long(
-                    non_empty_index.data_ptr(), non_empty_index.shape()
-                )  # tensor shape (-1), variable length
-
-                batch_size = len(sample_id_num)
-                dim = distinct_id_tensor.shape[-1]
-                sample_fixed_size = index_tensor.shape[1] // batch_size
-                index_select_raw_tensor = distinct_id_tensor.index_select(
-                    0, index_tensor.view(-1)
-                )
-                index_select_raw_tensor.requires_grad = self.is_training
-
-                raw_fixed_size_tensor = index_select_raw_tensor.view(
-                    -1, sample_fixed_size, dim
-                )
-                mask = (
-                    index_tensor.view(batch_size, sample_fixed_size, 1) != 0
-                ).half()  # generate mask
-                raw_fixed_size_tensor_with_mask = torch.cat(
-                    [raw_fixed_size_tensor, mask], dim=2
-                )
-                emb_tensors.append(
-                    (
-                        raw_embedding.name(),
-                        distinct_id_tensor,
-                        index_tensor,
-                        non_empty_index_tensor,
-                        index_select_raw_tensor,
-                    )
-                )
-                forward_tensors.append(raw_fixed_size_tensor_with_mask)
-            else:
-                emb = emb.get_sum_embedding()
-                batch.emb_slot.append([emb])
-
-                sum_tensor = pte.ptr_to_tensor_f16(
-                    emb.data_ptr(), emb.shape(), self.is_training
-                )
-                forward_tensors.append(sum_tensor)
-                emb_tensors.append((emb.name(), None, None, None, sum_tensor))
-
-        batch.forward_tensors = forward_tensors
-        batch.emb_tensors = emb_tensors
+        batch = super(TrainCtx, self).prepare_features(batch, is_training)
```

Comment:
`batch = super().prepare_features(batch, is_training)`

Reply:
Solved in https://github.com/PersiaML/PersiaML/pull/18/commits/44c21ffae2b6a3895503cc6bfd256a2d3d8645f0